Tutorial on ingesting and querying Theta sketches (#12723)
apache/druid: docs/tutorials/tutorial-sketches-theta.md

---
id: tutorial-sketches-theta
title: Approximations with Theta sketches
sidebar_label: Theta sketches
---

A common problem in clickstream analytics is counting unique things, like visitors or sessions. Generally this involves scanning through all detail data, because unique counts **do not add up** as you aggregate the numbers.

For instance, we might be interested in the number of visitors that watched episodes of a TV show. Let's say we found that at a given day, 1000 unique visitors watched the first episode, and 800 visitors watched the second episode. We may want to explore further trends, for example:

> - How many visitors watched _both_ episodes?

Author: Added it to line 55

- Theta sketches support **set operations**. Given two Theta sketches over subsets of the data, we can compute the union, intersection, or set difference of these two. This gives us the ability to answer the questions above about the number of visitors that watched a specific combination of episodes.

There is a lot of advanced math behind Theta sketches[^2]. But with Druid, you do not need to bother about the complex algorithms - Theta sketches just work!

[^2]: Specifically, the accuracy of the result is governed by the size _k_ of the Theta sketch, and by the operations you perform. See more details in the [Apache DataSketches documentation](https://datasketches.apache.org/docs/Theta/ThetaAccuracy.html). There's also a version of the sketch estimator `THETA_SKETCH_ESTIMATE_WITH_ERROR_BOUNDS` which takes an additional integer parameter and returns the error boundaries for the result in a JSON object.

This tutorial shows you how to create Theta sketches from your input data at ingestion time and how to run distinct count and set operation queries on the Theta sketches.

> For this tutorial, we'll assume you've already downloaded Druid as described in

Author: ```suggestion
## Prerequisites
For this tutorial, we'll assume you've already downloaded Druid as described in
```

![Configure schema for rollup and query granularity](../assets/tutorial-theta-03.png)

You also add the Theta sketch during this stage. Select **Add metric**.
Define the new metric as a Theta sketch with the following details:
* **Name**: `theta_uid`
* **Type**: `thetaSketch`
* **Field name**: `uid`
* **Size**: Leave at the default value, `16384`.
* **Is input theta sketch**: Leave at the default value, `False`.

![Create Theta sketch metric](../assets/tutorial-theta-04.png)

Click **Apply** to add the new metric to the data model.

We have to perform one more step to complete the data model. We are not interested in individual user ID's, only the unique counts. Right now, `uid` is still in the data model. Let's get rid of that!

Click on the `uid` column in the data model and delete it using the trashcan icon on the right:

![Delete uid column](../assets/tutorial-theta-05.png)

For the rest of the **Load data** wizard, set the following options:
* **Partition** stage: Set **Segment granularity** to `day`.
* **Tune**: Leave the default options.
* **Publish**: Set the datasource name to `ts_tutorial`.

> On the **Edit spec** page, your final input spec should look like the following JSON: 

Author: ```suggestion
On the **Edit spec** page, your final input spec should match the following:
```

```json
{
  "type": "index_parallel",
  "spec": {
    "ioConfig": {
      "type": "index_parallel",
      "inputSource": {
        "type": "inline",
        "data": "date,uid,show,episode\n2022-05-19,alice,Game of Thrones,S1E1\n2022-05-19,alice,Game of Thrones,S1E2\n2022-05-19,alice,Game of Thrones,S1E1\n2022-05-19,bob,Bridgerton,S1E1\n2022-05-20,alice,Game of Thrones,S1E1\n2022-05-20,carol,Bridgerton,S1E2\n2022-05-20,dan,Bridgerton,S1E1\n2022-05-21,alice,Game of Thrones,S1E1\n2022-05-21,carol,Bridgerton,S1E1\n2022-05-21,erin,Game of Thrones,S1E1\n2022-05-21,alice,Bridgerton,S1E1\n2022-05-22,bob,Game of Thrones,S1E1\n2022-05-22,bob,Bridgerton,S1E1\n2022-05-22,carol,Bridgerton,S1E2\n2022-05-22,bob,Bridgerton,S1E1\n2022-05-22,erin,Game of Thrones,S1E1\n2022-05-22,erin,Bridgerton,S1E2\n2022-05-23,erin,Game of Thrones,S1E1\n2022-05-23,alice,Game of Thrones,S1E1"
      },
      "inputFormat": {
        "type": "csv",
        "findColumnsFromHeader": true
      }
    },
    "tuningConfig": {
      "type": "index_parallel",
      "partitionsSpec": {
        "type": "hashed"
      },
      "forceGuaranteedRollup": true
    },
    "dataSchema": {

>       "dataSource": "inline_data",

Reviewer: should this be `ts_tutorial` to match line 187?

Reviewer: I didn't see anywhere in the Load data UI that you could set the datasource name for inline data, so there many need to be an instruction to set the `datasource` name or potentially change the ds name in the tutorial to match this.

Author: described in line 131

      "timestampSpec": {
        "column": "date",
        "format": "auto"
      },
      "dimensionsSpec": {
        "dimensions": [
          "show",
          "episode"
        ]
      },
      "granularitySpec": {
        "queryGranularity": "day",
        "rollup": true,
        "segmentGranularity": "day"
      },
      "metricsSpec": [
        {
          "name": "count",
          "type": "count"
        },
        {
          "type": "thetaSketch",
          "name": "theta_uid",
          "fieldName": "uid"
        }
      ]
    }
  }
}
```

Notice the `theta_uid` object in the `metricsSpec` list, which directs Druid to apply the `thetaSketch` aggregator on the `uid` column during ingestion.

Click **Submit** to start the ingestion.

## Query the Theta sketch column

Getting a unique count estimate out of a Theta sketch column involves two steps:

> 1. merging the Theta sketches in the column by means of an [aggregator function](../querying/sql-aggregations.md#theta-sketch-functions), which in Druid SQL is called `DS_THETA`

Reviewer: ```suggestion
1. Merge the Theta sketches in the column by means of the `DS_THETA` [aggregator function](../querying/sql-aggregations.md#theta-sketch-functions) in Druid SQL.
```

> 2. getting the estimate out of the merged sketch using [`THETA_SKETCH_ESTIMATE`](../querying/sql-scalar.md#theta-sketch-functions).

Reviewer: ```suggestion
2. Retrieve the estimate from the merged sketch with the [`THETA_SKETCH_ESTIMATE`](../querying/sql-scalar.md#theta-sketch-functions) function.
```

> Between steps 1 and 2, you can apply set functions. We will come to that in a moment.

Reviewer: ```suggestion
Between steps 1 and 2, you can apply set functions as demonstrated later? in section x? of the tutorial.

### Basic counting 

Let's first see what the data looks like in Druid. Run the following SQL statement in the query editor:
```sql

> SELECT * FROM ts_tutorial

Reviewer: The image shows 20 rows. There are only 19 rows in the input source. My results, with rollup, only shows 12 rows. Check to make sure this is accurate.

> SELECT * FROM ts_tutorial

Author: the horiz bar at the bottom configures the display but there are 12 rows shown next to the download icon

```

![View data with SELECT all query](../assets/tutorial-theta-06.png)

The Theta sketch column `theta_uid` appears as a Base64-encoded string; behind it is a bitmap.

> A simple query to compute distinct counts of user IDs uses `APPROX_COUNT_DISTINCT_DS_THETA` and groups by the other dimensions:

Reviewer: The following query to computes the distinct counts of user IDs uses `APPROX_COUNT_DISTINCT_DS_THETA` and groups by the other dimensions:
```

Author: ```suggestion
The following query to compute the distinct counts of user IDs uses `APPROX_COUNT_DISTINCT_DS_THETA` and groups by the other dimensions:
```

```sql
SELECT __time,
       "show",
       "episode",
       APPROX_COUNT_DISTINCT_DS_THETA(theta_uid) AS users
FROM   ts_tutorial
GROUP  BY 1, 2, 3
```

![Count distinct with Theta sketches](../assets/tutorial-theta-07.png)

In the query above, `APPROX_COUNT_DISTINCT_DS_THETA` is equivalent to calling `DS_THETA` and `THETA_SKETCH_ESIMATE` as follows:

```sql
SELECT __time,
       "show", 
       "episode",
       THETA_SKETCH_ESTIMATE(DS_THETA(theta_uid)) AS users
FROM   ts_tutorial
GROUP  BY 1, 2, 3

> ```

Reviewer: This only returns 12 rows for me, not 20

That is, `APPROX_COUNT_DISTINCT_DS_THETA` applies the following:
* `DS_THETA`: Creates a new Theta sketch from the column of Theta sketches
* `THETA_SKETCH_ESTIMATE`: Calculates the distinct count estimate from the output of `DS_THETA`

> 

Author: I think this query just demonstrates how to do the count distinct. Maybe it could be said that there are more unique watchers on 5/21 compared to days 5/20 and 5/19. If a user removes the `__time` grouping, you might be able to compare numbers.

### Filtered metrics

Druid has the capability to use [filtered metrics](../querying/sql-aggregations.md). This means we can include a WHERE clause in the SELECT part of the query.
> In the case of Theta sketches, the filter clause has to be inserted between the aggregator and the estimator.

As an example, query the total unique users that watched _Bridgerton:_

```sql
SELECT THETA_SKETCH_ESTIMATE(
         DS_THETA(theta_uid) FILTER(WHERE "show" = 'Bridgerton')
       ) AS users
FROM ts_tutorial
```

> 

Reviewer: This is returning empty for my data set which does not make sense

2022-05-22,bob,Game of Thrones,S1E1
2022-05-22,bob,Bridgerton,S1E1
2022-05-22,carol,Bridgerton,S1E2
2022-05-22,bob,Bridgerton,S1E1
2022-05-22,erin,Game of Thrones,S1E1
2022-05-22,erin,Bridgerton,S1E2
2022-05-23,erin,Game of Thrones,S1E1
2022-05-23,alice,Game of Thrones,S1E1
```

Navigate to the **Load data** wizard in the Druid console.
Select `Paste data` as the data source and paste the sample from above:

![Load data view with pasted data](../assets/tutorial-theta-01.png)

Leave the source type as `inline` and click **Apply** and **Next: Parse data**.
Parse the data as CSV, with included headers:

![Parse raw data](../assets/tutorial-theta-02.png)

Accept the default values in the **Parse time**, **Transform**, and **Filter** stages.

In the **Configure schema** stage, enable rollup and confirm your choice in the dialog. Then set the query granularity to `day`.

![Configure schema for rollup and query granularity](../assets/tutorial-theta-03.png)

You also add the Theta sketch during this stage. Select **Add metric**.
Define the new metric as a Theta sketch with the following details:
* **Name**: `theta_uid`
* **Type**: `thetaSketch`
* **Field name**: `uid`

> * **Size**: Leave at the default value, `16384`.
> * **Is input theta sketch**: Leave at the default value, `False`.

Author: ```suggestion
* **Size**: Accept the default value, `16384`.
* **Is input theta sketch**: Accept the default value, `False`.
```

sidebar_label: Theta sketches
---

<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one
  ~ or more contributor license agreements.  See the NOTICE file
  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ "License"); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

Apache Druid can power real-time collection, streaming, and interactive visualization of clickstreams.
A common problem in clickstream analytics is counting unique things, like visitors or sessions. Generally this involves scanning through all detail data, because unique counts do not add up as you aggregate the numbers.

> For instance, you might be interested in the number of visitors that watched episodes of a TV show. Let's say you found that at a given day, 1000 unique visitors watched the first episode, and 800 visitors watched the second episode. You may want to explore further trends, for example:

Reviewer: Consider adding a header before line 29 that introduces the problem scenario:
## The problem with performing set operations on large data sets ?

Then instead of "For instance, you might be interested..." you could say "Imagine you are interested..."

- How many visitors watched _both_ episodes?
- How many visitors are there that watched _at least one_ of the episodes?
- How many visitors watched episode 1 _but not_ episode 2?

There is no way to answer these questions by just looking at the aggregated numbers. You would have to go back to the detail data and scan every single row. If the data volume is high enough, this may take long, meaning that an interactive data exploration is not possible.

An additional nuisance is that unique counts don't work well with rollups. For this example, it would be great if you could have just one row of data per 15 minute interval[^1], show, and episode. After all, you are not interested in the individual user IDs, just the unique counts.

[^1]: Why 15 minutes and not just 1 hour? Intervals of 15 minutes work better with international timezones because those are not always aligned by hour. India, for instance, is 30 minutes off, and Nepal is even 45 minutes off. With 15 minute aggregates, you can get hourly sums for any of those timezones, too!

Is there a way to avoid crunching the detail data every single time, and maybe even enable rollup?

> **Enter Theta sketches for fast approximation with set operations.**

Reviewer: What if we end this section with "Enter Theta sketches"? 

nit: "enter" might confuse non-native English speaker, but not a major deal

Then on the next line
## Use Theta sketches for fast approximation with set operations

         )
       ) AS users
FROM ts_tutorial
```

![Count distinct with Theta sketches, filters, and set operations](../assets/tutorial-theta-10.png)

And finally, there is `THETA_SKETCH_NOT` which computes the set difference of two or more segments.
The result describes how many visitors watched episode 1 of Bridgerton but not episode 2.

```sql
SELECT THETA_SKETCH_ESTIMATE(
         THETA_SKETCH_NOT(
           DS_THETA(theta_uid) FILTER(WHERE "show" = 'Bridgerton' AND "episode" = 'S1E1'),
           DS_THETA(theta_uid) FILTER(WHERE "show" = 'Bridgerton' AND "episode" = 'S1E2')
         )
       ) AS users
FROM ts_tutorial
```

![Count distinct with Theta sketches, filters, and set operations](../assets/tutorial-theta-11.png)

## Conclusions

- Counting distinct things for large data sets can be done with Theta sketches in Apache Druid.
- This allows us to use rollup and discard the individual values, just retaining statistical approximations in the sketches.
- With Theta sketch set operations, affinity analysis is easier, for example, to answer questions such as which segments correlate or overlap by how much.

> ## Further reading

Reviewer: any reason we're not using "Learn more" ? 😬

Author: All the Druid tutorials that have this section call it "Further reading" so this keeps it in alignment

## Further reading

See the following topics for more information:
* [Theta sketch](../development/extensions-core/datasketches-theta.md) for reference on ingestion and native queries on Theta sketches in Druid.
* [Theta sketch scalar functions](../querying/sql-scalar.md#theta-sketch-functions) and [Theta sketch aggregation functions](../querying/sql-aggregations.md#theta-sketch-functions) for Theta sketch functions in Druid SQL queries.
* [Sketches for high cardinality columns](../ingestion/schema-design.md#sketches-for-high-cardinality-columns) for Druid schema design involving sketches.
* [DataSketches extension](../development/extensions-core/datasketches-extension.md) for more information about the DataSketches extension in Druid as well as other available sketches.
* The accuracy of queries using Theta sketches is governed by the size _k_ of the Theta sketch and by the operations you perform. See more details in the [Apache DataSketches documentation](https://datasketches.apache.org/docs/Theta/ThetaAccuracy.html).

## Acknowledgments

> This tutorial is adapted from a blog post by Hellmar Becker. Visit the [original blog post](https://blog.hellmar-becker.de/2022/06/05/druid-data-cookbook-counting-unique-visitors-for-overlapping-segments/) on Hellmar Becker's blog.

Reviewer 2: i've mixed opinions on this. On the one hand, it definitely seems a bit strange to me have this section in the docs that links out to another website, these docs are supposed to be the authority on Druid stuff if my mind. But on the other hand this is a pretty great content and it deserves credit.

Do we do stuff like this in other docs pages?

Do we do this anywhere else?

Reviewer: We haven't done this on other docs pages. However, @hellmarbecker is a member of the Druid community and the originator of this content. We have some way to acknowledge that his authorship. Perhaps this would be cleaner:

This tutorial is adapted from a [blog post by Hellmar Becker](https://blog.hellmar-becker.de/2022/06/05/druid-data-cookbook-counting-unique-visitors-for-overlapping-segments/).

it also cuts down some of the repetition.

Reviewer 2: I think the part that seemed strange is that it sort of seems vaguely like it is someone elses content, which isn't technically true if it is being (legally) contributed here.

Docs are contributed and licensed the same way as source files, and so this contribution is granting the ASF ownership of this content and licensing it to be distributed under the Apache license. The standard way we acknowledge external source contributions when necessary is the https://github.com/apache/druid/blob/master/LICENSE and https://github.com/apache/druid/blob/master/NOTICE files, though I'm unsure if it would be appropriate in this case, since the source packages are not the typical consumption method of the docs.

I just wanted to make sure that this is clear to @hellmarbecker to ensure that the license of this content is clean.

If all of this sounds ok, I do think it is ok to acknowledge and link to this external content since it was developed externally and later added instead of as an original Druid contribution (which we do not typically credit the author with, other than in the form of github commits and occasionally release announcements).

Reviewer 3: Thanks, I am happy to grant permission for Apache to use the content!
